### PR TITLE
Update tests to remove reliance on UserDefaults

### PR DESCRIPTION
**Pull request by aokj4ck**
### Description

- Remove all usages of UserDefaults from MapboxSearch and MapboxSearchUI
    - Replaces this with plain dependency injection
- Add LocalhostMockServiceProvider for integration tests to provide a mocked customBaseURL
- Add FeedbackIntegrationTestCase back to UI test suite
- Split MockServerTestCase into UI and Integration variants because it had a namespace collision
- Utilize customBaseURL for ServiceProvider and LocalhostMockServiceProvider
- Stop reading from  "MapboxAPIBaseURL" from NSUserDefaults in `ServiceProvider.createEngine`.
    - Providing a value in Info.plist is still supported.
- Update integration tests to use LocalhostMockServiceProvider
- This fixes a bug in the Demo app where you building and running the app or tests could use the wrong baseApiURL based on the current value in user defaults.
- Update and correct tests for iOS 17 using all mocked data for unit and UI tests.
- Rename `SearchEngine.reverseGeocoding` function to `SearchEngine.reverse`.
- OfflineTests are still failing and will be address later

### Checklist
- [x] Update `CHANGELOG`

### Screenshots

| Before | After |
| -- | -- |
| Screenshot 2024-02-06 at 17 47 49 | Screenshot 2024-02-06 at 17 28 38 |

### CODE REVIEW

File: Sources/MapboxSearch/PublicAPI/Engine/SearchEngine.swift [hunk 1]
```diff
@@ -490,6 +490,25 @@ extension SearchEngine {
...
+    /// Reverse geocoding of coordinates to addresses.
+    /// The default behavior in reverse geocoding is to return at most one feature at each of the multiple levels of the
+    /// administrative hierarchy (for example, one address, one region, one country).
+    /// Increasing the limit allows returning multiple features of the same type, but only for one type (for example,
+    /// multiple address results).
+    /// Consequently, setting limit to a higher-than-default value requires specifying exactly one types parameter.
+    /// - Parameters:
+    ///   - options: Options with coordinates, mode, limits and query types for reverse geocoding.
+    ///   - completion: completion handler with either reverse geocoding Resuts or Error.
+    @available(*, deprecated, renamed: "reverse")
```
Comment (azarovalex) [inline Sources/MapboxSearch/PublicAPI/Engine/SearchEngine.swift:505]: Considering that we already did a lot of breaking change in v2 and we are still doing prereleases, maybe we should just drop this old function?

File: Sources/MapboxSearchUI/MapboxSearchController.swift [hunk 2]
```diff
@@ -643,7 +643,7 @@ extension MapboxSearchController: SearchCategoriesRootViewDelegate {
             case .success(let items):
                 self.delegate?.categorySearchResultsReceived(category: category, results: items)
             case .failure(let searchError):
-                print("Failed search; error=\(searchError)")
+                print("Failed search; error=\(searchError)") // swiftlint:disable:this dont_use_print
```
Comment (azarovalex) [inline Sources/MapboxSearchUI/MapboxSearchController.swift:646]: We should be able to print through CommonSDK logger right?

File: Tests/MapboxSearchIntegrationTests/SearchEngineGeocodingIntegrationTests.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            if case .failure(let error) = result {
+                if case .reverseGeocodingFailed(let reasonError as NSError, _) = error {
+                    XCTAssert(reasonError.code == 500)
+                } else {
+                    XCTFail("Not expected")
+                }
+            } else {
+                XCTFail("Not expected")
+            }
+            expectation.fulfill()
```
Comment (azarovalex) [inline Tests/MapboxSearchIntegrationTests/SearchEngineGeocodingIntegrationTests.swift:56]: Just to simplify things a bit

```suggestion
        searchEngine.reverse(options: options) { result in
            if case .failure(.reverseGeocodingFailed(let reasonError as NSError, _)) = result {
                XCTAssert(reasonError.code == 500)
                expectation.fulfill()
            }
```

**Review by azarovalex — DISMISSED** [review]

File: Sources/MapboxSearch/PublicAPI/Engine/SearchEngine.swift (hunk 1 above)
Comment (aokj4ck) [reply Sources/MapboxSearch/PublicAPI/Engine/SearchEngine.swift:505]: Oh right good idea to drop it now.

**Review by azarovalex — APPROVED** [review]

**Pull request merged**